Install MMA Extension on windowsVMs

rupanter opened the pull request:
Added policy to install Microsoft Monitoring Agent(MMA) Extension on Windows Azure Virtual Machines based on tags.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-policy?pullRequest=435) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-policy?pullRequest=435) before we can accept your contribution.<br/><br/>:x: rupanter [sign now](https://cla.opensource.microsoft.com/Azure/azure-policy?pullRequest=435)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-policy?pullRequest=435) it.</sub>

rupanter commented:
@pilor - checking if you could review this please :)

pilor commented:
Please remove the .DS_Store file from this PR

pilor commented:
There is an existing Log Analytics agent policy here:  https://github.com/Azure/azure-policy/blob/master/samples/Compute/deploy-oms-vm-extension-windows-vm/azurepolicy.json

The only difference between the two is filtering based on the tag.  I'm OK with having both but they should be next to each other and reference each other in their READMEs.

rupanter commented:
@pilor I've removed the .DS_Store. Also the policy you are referring is using old terminology : OMS whereas now it is outdated and is changed to Log Analytics. I have moved the Policy under Monitoring and have changed oms to LA to reflect correctly. Regarding reference, One is image definition and other is Tag based, do we need to refer? If yes, are there any guidelines for it?

pilor commented on samples/Monitoring/deploy-la-vm-extension-windows-vm/azurepolicy.json:
```diff
@@ -0,0 +1,129 @@
+                "metadata": {
+                    "displayName": "Log Analytics workspace",
+                    "description": "Select Log Analytics workspace from dropdown list",
+                    "strongType": "LAWorkspace"
```
>LAWorkspace [](start = 35, length = 11)

The strongType is still "omwWorkspace"

pilor commented on samples/Monitoring/deploy-la-vm-extension-windows-vm/azurepolicy.parameters.json:
```diff
@@ -0,0 +1,10 @@
+        "metadata": {
+            "displayName": "Log Analytics workspace",
+            "description": "Select Log Analytics workspace from dropdown list",
+            "strongType": "LAWorkspace"
```
>LAWorkspace [](start = 27, length = 11)

omsWorkspace

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+{
+    "properties": {
+        "displayName": "Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag",
```
> Microsoft Monitoring Agent [](start = 30, length = 27)

please change all references to MMA to "Log Analytics Agent"

pilor replied:
The correct terminology according to the Log Analytics team is "Log Analytics Agent".  It is what they use in their built-in policies and documentation.

---
In reply to: [310304741](https://github.com/Azure/azure-policy/pull/435#discussion_r310304741) [](ancestors = 310304741)

rupanter replied:
I've corrected this - Agent name to be Log Analytics Agent.

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+{
+    "properties": {
+        "displayName": "Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag",
```
>Machines [](start = 69, length = 8)

VMs

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+        "mode": "Indexed",
+        "description": "Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag with value: MMA and Name: Windows",
+        "metadata": {
+            "category": "Batch"
```
>Batch [](start = 25, length = 5)

Monitoring

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+    "properties": {
+        "displayName": "Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag",
+        "mode": "Indexed",
+        "description": "Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag with value: MMA and Name: Windows",
```
> with value: MMA and Name: Windows [](start = 92, length = 34)

The user gets to specify the tag key in the policy params

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+                "metadata": {
+                    "displayName": "Log Analytics workspace",
+                    "description": "Name of the workspace, such as 'production'"
+                }
```
This isn't accurate.  The WorkspaceId needs to be the .customerId from the workspace in order for this policy to work... Why not get it via reference() like the built-in policy?

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+                "metadata": {
+                    "displayName": "Tag Name",
+                    "description": "Name of the tag, such as 'production'"
+                }
```
needs more description details how the tag is used

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+                            {
+                                "field": "Microsoft.Compute/virtualMachines/extensions/provisioningState",
+                                "equals": "Succeeded"
+                            }
```
I know the built-in does this but this typically isn't a good idea.  If a user-provided LA extension isn't done deploying or is in a bad state this will add a new LA extension with the name specified in this policy.  The policy should only add an LA extension if one does not exist.  If you want to audit failed LA extensions that should be its own policy

rupanter replied:
I understand this and but since the built-in does this hence I went with the same approach. I will come up with an Audit policy as well. Thanks

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/README.md:
```diff
@@ -0,0 +1,30 @@
+
+az policy definition create --name 'Deploy-mmaextension-windows-tags-loganalytics' --display-name 'Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag' --description 'Deploy Microsoft Monitoring Agent on Windows Machines based on a Tag' --rules 'https://raw.githubusercontent.com/Azure/azure-policy/master/samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy ...
+
+az policy assignment create --name <assignmentname> --scope <scope> --params "{'metricName':{'value':'<metricName>'}}" --policy "deploy-mmaextension-windows-tags-loganalytics"
```
please test the commands you are using here.  They won't work as written

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+    "properties": {
+        "displayName": "Deploy Microsoft Monitoring Agent on Windows VMs based on a Tag",
+        "mode": "Indexed",
+        "description": "Deploy Microsoft Monitoring Agent on Windows VMs based on a Tag with value: MMA and Name: Windows",
```
>Tag with value: MMA and Name: Windows [](start = 84, length = 37)

this is still not accurate.  User provides the tag name/value as parameters

rupanter replied:
I missed this, I've updated it.

pilor commented on samples/Monitoring/deploy-mmaextension-windows-tags-loganalytics/azurepolicy.json:
```diff
@@ -0,0 +1,140 @@
+                                            "stopOnMultipleConnections": "true"
+                                        },
+                                        "protectedSettings": {
+                                            "workspaceKey": "[listKeys(resourceId('Microsoft.OperationalInsights/workspaces', parameters('loganalytics')), '2015-03-20').primarySharedKey]"
```
>listKeys(resourceId('Microsoft.OperationalInsights/workspaces', parameters('loganalytics') [](start = 62, length = 90)

this also needs to be updated

rupanter replied:
I have tested this locally and it works as expected. I am using resourceId instead of reference here and i've tested it :
![image](https://user-images.githubusercontent.com/22794646/63453506-70d18b80-c466-11e9-8d7c-8881737d38ad.png)

pilor replied:
This will be a runtime problem during template deployment execution, not a problem with the policy itself.  the resourceId() function takes type, name, etc... separately.  The "loganalytics" parameter is the full log analytics workspace resource ID.  You should be using it here very similarly to how it was used in the "reference()" function above.

[https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-resource#resourceid](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-resource#resourceid)

---
In reply to: [316305304](https://github.com/Azure/azure-policy/pull/435#discussion_r316305304) [](ancestors = 316305304)

rupanter replied:
> This will be a runtime problem during template deployment execution, not a problem with the policy itself. the resourceId() function takes type, name, etc... separately. The "loganalytics" parameter is the full log analytics workspace resource ID. You should be using it here very similarly to how it was used in the "reference()" function above.
> 
> https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-resource#resourceid
> 
> In reply to: [316305304](https://github.com/Azure/azure-policy/pull/435#discussion_r316305304) [](ancestors = 316305304)

I have made the changes - good call outs @pilor

pilor replied:
The deployment in your tags policy will still fail.  There are parameters in the deployment that aren't provided, the resourceID function in listKeys is invalid, etc...

kenieva commented:
All sample policy definitions pull requests should go to our community repo: https://github.com/Azure/CommunityPolicyRepo. Thank you!
